### ISSUE TITLE
Regarding server performance with LoRA

### ISSUE THREAD

**Issue by binhtranmcs** [NVIDIA/TensorRT-LLM#2441]
### System Info

GPU: Nvidia H100
Model: Llama3 8B

### Who can help?

@kaiyux 

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

I follow the steps [here](https://github.com/NVIDIA/TensorRT-LLM/tree/main/benchmarks/cpp) to prepare dataset and benchmark server with multiple lora adapters:

```
# Generate random dataset
python benchmarks/cpp/prepare_dataset.py \
  --stdout \
  --tokenizer=$HF_DIR \
  token-norm-dist \
  --num-requests=1000 \
  --input-mean=128 --output-mean=128 \
  --input-stdev=0 --output-stdev=0 > $BENCHMARK_DIR/token-norm-dist.json

# Generate random dataset with 3 lora adapters
python benchmarks/cpp/prepare_dataset.py \
  --output=$BENCHMARK_DIR/token-norm-dist-lora-3.json \
  --rand-task-id 1 3 \
  --tokenizer=$HF_DIR \
  token-norm-dist \
  --num-requests=1000 \
  --input-mean=128 --output-mean=128 \
  --input-stdev=0 --output-stdev=0 > $BENCHMARK_DIR/token-norm-dist-lora-3.json

# Generate random lora weights for 3 adapters
python benchmarks/cpp/utils/generate_rand_loras.py \
  $LORA_DIR \
  $LORA_TEST_DIR 4

# Perform benchmarking

# First run inference without lora
./gptManagerBenchmark \
  --engine_dir $ENGINE_DIR \
  --request_rate 300 \
  --dataset $BENCHMARK_DIR/token-norm-dist.json \
  --max_num_samples 1000 \
  --kv_cache_free_gpu_mem_fraction 0.9 \
  --enable_kv_cache_reuse \
  --enable_chunked_context \
  --eos_id -100

# Now run inference with various numbers or loras
./gptManagerBenchmark \
  --engine_dir $ENGINE_DIR \
  --request_rate 300 \
  --dataset $BENCHMARK_DIR/token-norm-dist-lora-3.json \
  --max_num_samples 1000 \
  --kv_cache_free_gpu_mem_fraction 0.9 \
  --enable_kv_cache_reuse \
  --enable_chunked_context \
  --eos_id -100 \
  --lora_dir $LORA_TEST_DIR
```

I tested with multiple max batch size and got the results:

- Max batch size 2048:
```
- 0 lora:
[BENCHMARK] seq_throughput(seq/sec) 164.49
[BENCHMARK] token_throughput(token/sec) 21054.79
- 1 lora:
[BENCHMARK] seq_throughput(seq/sec) 113.86
[BENCHMARK] token_throughput(token/sec) 14574.02
- 3 lora:
[BENCHMARK] seq_throughput(seq/sec) 51.60
[BENCHMARK] token_throughput(token/sec) 6605.11
```
- Max batch size 128:
```
- 0 lora:
[BENCHMARK] seq_throughput(seq/sec) 104.71
[BENCHMARK] token_throughput(token/sec) 13402.54
- 1 lora:
[BENCHMARK] seq_throughput(seq/sec) 57.36
[BENCHMARK] token_throughput(token/sec) 7342.44
- 3 lora:
[BENCHMARK] seq_throughput(seq/sec) 39.74
[BENCHMARK] token_throughput(token/sec) 5086.20
```
- Max batch size 32:
```
- 0 lora:
[BENCHMARK] seq_throughput(seq/sec) 39.80
[BENCHMARK] token_throughput(token/sec) 5094.89
- 1 lora:
[BENCHMARK] seq_throughput(seq/sec) 19.84
[BENCHMARK] token_throughput(token/sec) 2539.42
- 3 lora:
[BENCHMARK] seq_throughput(seq/sec) 18.55
[BENCHMARK] token_throughput(token/sec) 2374.53
```

### Expected behavior

Throughput does not drop so much when deploying with LoRA.

### actual behavior

Throughput dropped by nearly half with just 1 LoRA.

### additional notes

You can see the performance is significantly worse when using even just 1 LoRA. I wonder if you have a benchmark report when using TensorRT-LLM with multiple LoRAs? Also do you have any suggestion on why the throughput dropped so much? Thanks in advance.

**Comment by avianion** [comment]
@binhtranmcs i am interested to know your results. can you try with 100 loras?

**Comment by byshiue** [comment]
Can you share the full scripts to reproduce the issue? e.g., what kind of HF model do you use? What kind of lora model do you use? It would be helpful to reproduce the issue if you can share the full scripts.

**Comment by binhtranmcs** [comment]
@byshiue here are the steps to reproduce:
```
# Convert checkpoint
python3 examples/llama/convert_checkpoint.py \
  --workers 8 --model_dir meta-llama/Meta-Llama-3-8B-Instruct \
  --output_dir $CONVERT_DIR \
  --dtype float16 --tp_size 1 --pp_size 1

# Build engine
trtllm-build --workers 8 \
  --checkpoint_dir=$CONVERT_DIR \
  --output_dir=$ENGINE_DIR \
  --gemm_plugin float16 --gpt_attention_plugin float16 --use_paged_context_fmha enable \
  --lora_plugin float16 \
  --lora_target_modules attn_q attn_k attn_v attn_dense mlp_gate mlp_h_to_4h mlp_4h_to_h \
  --max_lora_rank 64 \
  --kv_cache_type paged --remove_input_padding enable --context_fmha enable \
  --max_batch_size 128 --max_beam_width 1 --max_input_len 128 --max_seq_len 256 --logits_dtype float16

# Generate random dataset
python benchmarks/cpp/prepare_dataset.py \
  --stdout \
  --tokenizer=$HF_DIR \
  token-norm-dist \
  --num-requests=1000 \
  --input-mean=128 --output-mean=128 \
  --input-stdev=0 --output-stdev=0 > $BENCHMARK_DIR/token-norm-dist.json

# Generate random dataset with 3 lora adapters
python benchmarks/cpp/prepare_dataset.py \
  --output=$BENCHMARK_DIR/token-norm-dist-lora-3.json \
  --rand-task-id 1 3 \
  --tokenizer=$HF_DIR \
  token-norm-dist \
  --num-requests=1000 \
  --input-mean=128 --output-mean=128 \
  --input-stdev=0 --output-stdev=0 > $BENCHMARK_DIR/token-norm-dist-lora-3.json

# Generate random lora weights for 3 adapters
python benchmarks/cpp/utils/generate_rand_loras.py \
  $LORA_DIR \
  $LORA_TEST_DIR 4

### Perform benchmarking

# First run inference without lora
./gptManagerBenchmark \
  --engine_dir $ENGINE_DIR \
  --request_rate 300 \
  --dataset $BENCHMARK_DIR/token-norm-dist.json \
  --max_num_samples 1000 \
  --kv_cache_free_gpu_mem_fraction 0.9 \
  --enable_kv_cache_reuse \
  --enable_chunked_context \
  --eos_id -100

# Now run inference with various numbers or loras
./gptManagerBenchmark \
  --engine_dir $ENGINE_DIR \
  --request_rate 300 \
  --dataset $BENCHMARK_DIR/token-norm-dist-lora-3.json \
  --max_num_samples 1000 \
  --kv_cache_free_gpu_mem_fraction 0.9 \
  --enable_kv_cache_reuse \
  --enable_chunked_context \
  --eos_id -100 \
  --lora_dir $LORA_TEST_DIR
```

The example lora is [here](https://drive.google.com/file/d/1ZSO8Vhz7xCRgONy3uBx8BY0j78fUA7OQ/view?usp=sharing).
Please have a look.

**Comment by byshiue** [comment]
> [@byshiue](https://github.com/byshiue) here are the steps to reproduce:
> 
> ```
> # Convert checkpoint
> python3 examples/llama/convert_checkpoint.py \
>   --workers 8 --model_dir meta-llama/Meta-Llama-3-8B-Instruct \
>   --output_dir $CONVERT_DIR \
>   --dtype float16 --tp_size 1 --pp_size 1
> 
> # Build engine
> trtllm-build --workers 8 \
>   --checkpoint_dir=$CONVERT_DIR \
>   --output_dir=$ENGINE_DIR \
>   --gemm_plugin float16 --gpt_attention_plugin float16 --use_paged_context_fmha enable \
>   --lora_plugin float16 \
>   --lora_target_modules attn_q attn_k attn_v attn_dense mlp_gate mlp_h_to_4h mlp_4h_to_h \
>   --max_lora_rank 64 \
>   --kv_cache_type paged --remove_input_padding enable --context_fmha enable \
>   --max_batch_size 128 --max_beam_width 1 --max_input_len 128 --max_seq_len 256 --logits_dtype float16
> 
> # Generate random dataset
> python benchmarks/cpp/prepare_dataset.py \
>   --stdout \
>   --tokenizer=$HF_DIR \
>   token-norm-dist \
>   --num-requests=1000 \
>   --input-mean=128 --output-mean=128 \
>   --input-stdev=0 --output-stdev=0 > $BENCHMARK_DIR/token-norm-dist.json
> 
> # Generate random dataset with 3 lora adapters
> python benchmarks/cpp/prepare_dataset.py \
>   --output=$BENCHMARK_DIR/token-norm-dist-lora-3.json \
>   --rand-task-id 1 3 \
>   --tokenizer=$HF_DIR \
>   token-norm-dist \
>   --num-requests=1000 \
>   --input-mean=128 --output-mean=128 \
>   --input-stdev=0 --output-stdev=0 > $BENCHMARK_DIR/token-norm-dist-lora-3.json
> 
> # Generate random lora weights for 3 adapters
> python benchmarks/cpp/utils/generate_rand_loras.py \
>   $LORA_DIR \
>   $LORA_TEST_DIR 4
> 
> ### Perform benchmarking
> 
> # First run inference without lora
> ./gptManagerBenchmark \
>   --engine_dir $ENGINE_DIR \
>   --request_rate 300 \
>   --dataset $BENCHMARK_DIR/token-norm-dist.json \
>   --max_num_samples 1000 \
>   --kv_cache_free_gpu_mem_fraction 0.9 \
>   --enable_kv_cache_reuse \
>   --enable_chunked_context \
>   --eos_id -100
> 
> # Now run inference with various numbers or loras
> ./gptManagerBenchmark \
>   --engine_dir $ENGINE_DIR \
>   --request_rate 300 \
>   --dataset $BENCHMARK_DIR/token-norm-dist-lora-3.json \
>   --max_num_samples 1000 \
>   --kv_cache_free_gpu_mem_fraction 0.9 \
>   --enable_kv_cache_reuse \
>   --enable_chunked_context \
>   --eos_id -100 \
>   --lora_dir $LORA_TEST_DIR
> ```
> 
> The example lora is [here](https://drive.google.com/file/d/1ZSO8Vhz7xCRgONy3uBx8BY0j78fUA7OQ/view?usp=sharing). Please have a look.

Can you prevent using the environment variables without defining them? For example, I don't know what `LORA_TEST_DIR` and `LORA_DIR` should be.

**Comment by binhtranmcs** [comment]
@byshiue `LORA_DIR` is the path to the lora adapter I attatched above. It will generate random loras and place them in `LORA_TEST_DIR`. Others environment variables are just paths you want to use to store generated files. You can refer to [this](https://github.com/NVIDIA/TensorRT-LLM/tree/main/benchmarks/cpp#benchmarking-lora) if it makes more sense.

**Comment by crossxxd** [comment]
The same problem occurred. I've encountered the same performance issue as that in the https://github.com/NVIDIA/TensorRT-LLM/issues/1852. I used v0.14 and Tritonserver 24.10, and it seems that the problem hasn't been optimized and solved as replied in issue 1852.

**Comment by byshiue** [comment]
@crossxxd The issue is closed automatically because it is staled.

@binhtranmcs Thank you for reporting. It would be helpful if you can prevent something like **Others environment variables are just paths you want to use to store generated files**. Guessing what kind of variable should we use spends lots of time. It would be simpler to reproduce your issue and give the reply ASAP if you can share your full scripts directly.

I run some benchmarks on batch size 128

|         -                                       | no lora | no lora with lora module | 1 lora | 1 lora, sorted | 3 lora | 3 lora, sorted | 3 lora, window size 16|
|:---------------------------------:|:--------:|:---------------------------:|:-------:|:--------------:|:-------:|:---------------:|:--------------------------:|
| token throughput (token/sec) |  6586    | 5404   (82%)                    | 4198 (63%)  |  4900 (74%)   |  3518 (53%)  |  4716 (71%)  |  4552 (69%) |          

* **no lora**: Build the engine without lora modules
* **no lora with lora module**: Build the engine with lora modules, but not pass lora pointers. The overheads are caused by initializing the lora results and combine the lora results with GEMMs.
* **1 lora**: Only possible to pass 1 lora weights. So, the task ids might be -1 (not use lora) or 0 (use lora task id 0)
* **1 lora, sorted**: Simlar to **1 lora**, but sort the requests by task ids. So, when we have 1000 samples, the task ids of first 500 samples are all -1, and the task ids of last 500 samples are all 0. 
* **3 lora**: lora task ids are between [-1, 2]. 
* **3 lora, sorted**: Similar to **3 lora**, but sort the requests by task ids.
* **3 lora, window size 16**: Similar to **3 lora**, but partially sort the requests by task ids. Every 16 requests have same lora task ids. For example, the task ids of the first 16 samples are all 0, and the task ids of samples 17 ~ 32 are all 1, ... 


From the benchmark, we can observe the obvious performance drop if we don't do any preprocessing for the requests (**1 lora** and **3 lora** cases). That's because the group gemms are very small (batch size are often 1). But I don't fully reproduce the performance you share (drop more than 50%). 

To achieve better performance, we need to partially sort the request to enahcne the utilization of GPU. The **1 lora, sorted** and **3 lora, sorted** are the ideal cases in current implementation. Compared to **no lora with lora module**, the token throughput drops about 10%. And the **3 lora, window size 16** is a possible real case. If the server could try to collect the requests with same lora ids together, we can get better performance. 

We have following tasks in our plan:
1. Optimize the **no lora with lora module** case to prevent such large overhead even if we don't really need to use lora. 
2. Consider how to collect the requests with same lora task ids on our scheduler to improve the utilization of GPU usage.

Finally, thank you for reporting this issue again. It is helpful.
